tools/schemacode/schemacode/tests/test_validator.py

```diff
@@ -221,6 +221,47 @@ def test__add_suffixes():
     assert _regex_string == regex_string
 
 
+def test__inheritance_expansion():
+    from schemacode.validator import _inheritance_expansion
+
+    base_entry = (
+        r".*?/sub-(?P<subject>([0-9a-zA-Z]+))/"
+        r"(|ses-(?P<session>([0-9a-zA-Z]+))/)func/sub-(?P=subject)"
+        r"(|_ses-(?P=session))_task-(?P<task>([0-9a-zA-Z]+))"
+        r"(|_acq-(?P<acquisition>([0-9a-zA-Z]+)))"
+        r"(|_ce-(?P<ceagent>([0-9a-zA-Z]+)))"
+        r"(|_rec-(?P<reconstruction>([0-9a-zA-Z]+)))"
+        r"(|_dir-(?P<direction>([0-9a-zA-Z]+)))"
+        r"(|_run-(?P<run>([0-9]*[1-9]+[0-9]*)))"
+        r"(|_echo-(?P<echo>([0-9]*[1-9]+[0-9]*)))"
+        r"_phase(\.nii\.gz|\.nii|\.json)$"
+    )
+    expected_entries = [
+        ".*?/sub-(?P<subject>([0-9a-zA-Z]+))/"
+        "(|ses-(?P<session>([0-9a-zA-Z]+))/)"
+        "sub-(?P=subject)(|_ses-(?P=session))"
+        "_task-(?P<task>([0-9a-zA-Z]+))"
+        "(|_acq-(?P<acquisition>([0-9a-zA-Z]+)))"
+        "(|_ce-(?P<ceagent>([0-9a-zA-Z]+)))"
+        "(|_rec-(?P<reconstruction>([0-9a-zA-Z]+)))"
+        "(|_dir-(?P<direction>([0-9a-zA-Z]+)))"
+        "(|_run-(?P<run>([0-9]*[1-9]+[0-9]*)))"
+        "(|_echo-(?P<echo>([0-9]*[1-9]+[0-9]*)))"
+        "_phase(\\.nii\\.gz|\\.nii|\\.json)$",
+        ".*?/task-(?P<task>([0-9a-zA-Z]+))"
+        "(|_acq-(?P<acquisition>([0-9a-zA-Z]+)))"
+        "(|_ce-(?P<ceagent>([0-9a-zA-Z]+)))"
+        "(|_rec-(?P<reconstruction>([0-9a-zA-Z]+)))"
+        "(|_dir-(?P<direction>([0-9a-zA-Z]+)))"
+        "(|_run-(?P<run>([0-9]*[1-9]+[0-9]*)))"
+        "(|_echo-(?P<echo>([0-9]*[1-9]+[0-9]*)))"
+        "_phase(\\.nii\\.gz|\\.nii|\\.json)$",
+    ]
+
+    inheritance_expanded_entries = _inheritance_expansion(base_entry, datatype="func")
+    assert inheritance_expanded_entries == expected_entries
+
+
```

Comment:
I think it would have been nice if there were somewhat more straightforward tests on some listings of files (like the ones in the spec at https://bids-specification.readthedocs.io/en/latest/02-common-principles.html#the-inheritance-principle) which would check what files would be considered ok (or not ok) through "inheritance" principle. E.g. those from bids specification:

ok:
- `task-rest_bold.json`
- `sub-01/ses-test/sub-01_ses-test_task-overtverbgeneration_bold.json`

not ok (I guess ;)):
- `task-rest.json`
- `sub-01/sub-01_ses-test_task-overtverbgeneration_bold.json`
- `sub-01/ses-test/task-overtverbgeneration_bold.json`
- ... become creative ... ;)

Reply:
It might seem more straightforward on the surface, but we're not creating file names, rather regex patterns. This could be done, but it would mean an additional function to check whether two filenames match the different regex patterns created, it would be a util function for the tests, not actually used by the implementation, and I would argue it would make the inspection more intransparent.

If you're concerned about testing for false positives in inheritance validation, this would be better tested and verify the actual validation code if we had a dataset with the broken inheritance filenames you listed above. We can add that to our new bids-error-examples repo, I made an issue: https://github.com/bids-standard/bids-error-examples/issues/1

Reply:
as long as this particular validations are operating on filenames only, it is IMHO much easier/lightweight just to create test cases in the code which populate list of files in bids dataset which are given to the functions to be tested.  It makes it easier and faster then to create parametric tests etc, instead of relying on instantiation in real datasets as in a repo.

Reply:
> faster then [...] relying on instantiation in real datasets
this is true, but with regard to that we can simply optimize download so that we only fetch what we need: https://github.com/bids-standard/bids-specification/pull/1143

The problem is that the only valid reference for “true” BIDS data are the bids repositories, everything else can get out of sync under the radar.

We support validation for a list of paths, and we test that here: https://github.com/bids-standard/bids-specification/blob/9fd2a9951c45a9038079be257c44f6751ebf5e45/tools/schemacode/schemacode/tests/test_validator.py#L332
but note that this also generates the file list from the bids-examples repo.

In as far as we have hard-coded test input, this is for unit tests where we need the specific strings in the absence of the loaders which generate them, this is the case for testing this function, and we provide the regexes it operates on. Working with real filenames is more of an integration test case... We could do it if we really want to, but again, this would be a wrapper function which applies the patterns to hard-coded values, which might cause both false positives and false negatives as it would be additional logic used nowhere else but in the tests.

Reply:
if you insist :)